lsp-mode.el

```diff
@@ -5322,6 +5328,42 @@ argument ask the user to select which language server to start. "
                (apply 'concat (--map (format "[%s]" (lsp--workspace-print it))
                                      lsp--buffer-workspaces)))))
 
+(defvar lsp--emacs-started nil
+  "A boolean value indicating whether Emacs has finished starting.")
+
+(add-hook 'emacs-startup-hook #'lsp--emacs-startup)
```

Comment:
can this be part of use-package configuration? AFAICS if this hook is attached after emacs has started lsp--emacs-started will never be set to t. Can we document that if you want to use this functionality you will have to eagerly load lsp-mode on startup or you should manually call `lsp--emacs-startup` in your config (if this is the case it should be public)?

Reply:
Yes, you are right of course. I had some stupid config that was always loading lsp-mode. Anyway I found that it is desktop-mode itself that calls `switch-to-buffer` for each buffer as it is restoring the desktop file. That triggers `window-configuration-change-hook` and gives us false positives during startup.

I'd like to find some way to detect this so users don't have to configure anything other than `lsp-deferred`. Can you think of a good way to tell the difference between `lsp-deferred` being called via desktop-read and other cases? So far my only idea is to add a timer in `lsp--init-if-visible` so it only calls `lsp` if the buffer is still visible after 0.1 seconds (but I don't really like that idea).

Reply:
I checked the code in startup.el and it seems like we may use `inhibit-startup-hooks` to check whether the startup hook has already been run.

Reply:
I noticed `inhibit-startup-hooks`, too, but it doesn't actually get set for me after starting up.

Reply:
I think I found a better solution. Setting a timer for 0 seconds seems to avoid the false positives from desktop mode.

Reply:
Then the only solution that I could thick of is to ask the users to use emacs-startup-hook when lsp-mode is loaded on startup but call the lsp-mode `lsp--emacs-startup` explicitly if lsp-mode is lazy loaded.